src/ATokenCheckpointed.sol

```diff
@@ -0,0 +1,307 @@
+
+    // We need 256 bits because of the multiplication we're about to do.
+    uint256 _votingWeightAtSnapshot = IVotingToken(address(_underlyingAsset)).getPastVotes(
+      address(this), _proposalSnapshotBlockNumber
+    );
+
+    //      forVotesRaw          forVotesScaled
+    // --------------------- = ---------------------
+    //     totalDeposits        deposits (@snapshot)
+    //
+    // forVotesScaled = forVotesRaw * deposits@snapshot / totalDeposits
+    uint128 _forVotesToCast = SafeCast.toUint128(
+      (_votingWeightAtSnapshot * _proposalVote.forVotes) / _totalDepositWeightAtSnapshot
+    );
+    uint128 _againstVotesToCast = SafeCast.toUint128(
+      (_votingWeightAtSnapshot * _proposalVote.againstVotes) / _totalDepositWeightAtSnapshot
+    );
+    uint128 _abstainVotesToCast = SafeCast.toUint128(
+      (_votingWeightAtSnapshot * _proposalVote.abstainVotes) / _totalDepositWeightAtSnapshot
+    );
+
+    // This param is ignored by the governor when voting with fractional
+    // weights. It makes no difference what vote type this is.
+    uint8 unusedSupportParam = uint8(VoteType.Abstain);
+
+    hasCastVotesOnProposal[proposalId] = true;
+    bytes memory fractionalizedVotes =
+      abi.encodePacked(_forVotesToCast, _againstVotesToCast, _abstainVotesToCast);
+    GOVERNOR.castVoteWithReasonAndParams(
+      proposalId, unusedSupportParam, "crowd-sourced vote", fractionalizedVotes
+    );
+  }
+
+  /// @notice Implements the basic logic to mint a scaled balance token.
+  /// @param caller The address performing the mint
+  /// @param onBehalfOf The address of the user that will receive the scaled tokens
+  /// @param amount The amount of tokens getting minted
+  /// @param index The next liquidity index of the reserve
+  /// @return `true` if the the previous balance of the user was 0
+  function _mintScaledWithCheckpoint(
+    address caller,
+    address onBehalfOf,
+    uint256 amount,
+    uint256 index
+  ) internal returns (bool) {
+    bool _mintScaledReturn = _mintScaled(caller, onBehalfOf, amount, index);
+
+    (uint256 _preMintRawBalance, uint256 _postMintRawBalance) = _checkpointRawBalanceOf(onBehalfOf);
+    totalDepositCheckpoints.push(
+      totalDepositCheckpoints.latest() + (_postMintRawBalance - _preMintRawBalance)
+    );
+
+    return _mintScaledReturn;
+  }
+
+  function _rawBalanceOf(address _user) internal returns (uint256) {
+    return _userState[_user].balance;
+  }
+
+  function _checkpointRawBalanceOf(address _user)
+    internal
+    returns (uint256 _previousBalance, uint256 _currentBalance)
+  {
+    (_previousBalance, _currentBalance) = balanceCheckpoints[_user].push(_rawBalanceOf(_user));
+  }
+
+  // Returns the raw (i.e. unrebased) balanceOf the _user at the _blockNumber.
+  function getPastStoredBalance(address _user, uint256 _blockNumber) public returns (uint256) {
+    // TODO could we use getAtProbablyRecentBlock instead?
```

Comment:
docs for reference: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/Checkpoints.sol#L43-L48

I think `getAtProbablyRecentBlock` seems more appropriate 👍

Reply:
Cool, thanks! I'll make the change

Reply:
Done in 9d8171bb0a91c502d19cd1fb1180a7440f578c7f